runtime/caml/platform.h

```diff
@@ -120,6 +92,231 @@ void caml_plat_broadcast(caml_plat_cond*);
+/* Check if the barrier has been released */
+Caml_inline int caml_plat_barrier_is_released(caml_plat_barrier* barrier) {
+  return atomic_load_acquire(&barrier->futex.value) == Barrier_released;
+}
+
+/* Release and wait, but on a futex only.
+
+   This is like a(n inverted) binary semaphore, but with no decrement
+   on `wait`. That is, `release` sets the futex to 0, which `wait`
+   waits for.
+
+   A futex used this way is 0 (Barrier_released) when released, and
+   nonzero otherwise. It should be set to 1 (Barrier_unreleased) to
+   block.
+*/
+void caml_plat_barrier_raw_release(caml_plat_futex* futex);
+void caml_plat_barrier_raw_wait(caml_plat_futex* futex);
+
+/* Release the barrier unconditionally, letting all parties through */
+Caml_inline void caml_plat_barrier_release(caml_plat_barrier* barrier) {
+  caml_plat_barrier_raw_release(&barrier->futex);
+}
+/* Block until released */
+Caml_inline void caml_plat_barrier_wait(caml_plat_barrier* barrier) {
+  caml_plat_barrier_raw_wait(&barrier->futex);
+}
+
+/* -- Sense-reversing -- */
+/* Flip the sense of the barrier, releasing current waiters and
+   blocking new ones.
+
+   `current_sense` should be just `(b & BARRIER_SENSE_BIT)`
+   with b as returned by `caml_plat_barrier_arrive`. */
+void caml_plat_barrier_flip(caml_plat_barrier*, barrier_status current_sense);
+Caml_inline int
+caml_plat_barrier_sense_has_flipped(caml_plat_barrier* barrier,
+                                    barrier_status current_sense)
+{
+  return (atomic_load_acquire(&barrier->futex.value) & BARRIER_SENSE_BIT)
+    != current_sense;
+}
+/* Block until flipped */
+void caml_plat_barrier_wait_sense(caml_plat_barrier*,
+                                  barrier_status current_sense);
+
+/* Spin-wait loops */
+
+#define GENSYM_3(name, l) name##l
+#define GENSYM_2(name, l) GENSYM_3(name, l)
+#define GENSYM(name) GENSYM_2(name, __LINE__)
+
+#define Max_spins_long 1000
+#define Max_spins_medium 300
+#define Max_spins_short 30
+
+/* Spin up to some number of times, should be used when we have useful
+   work to do, or expect the condition to come true fast */
+#define SPIN_WAIT_BOUNDED SPIN_WAIT_NTIMES(Max_spins_medium)
+#define SPIN_WAIT_BOUNDED_LONG SPIN_WAIT_NTIMES(Max_spins_long)
+#define SPIN_WAIT_NTIMES(N)                             \
+  unsigned GENSYM(caml__spins) = 0;                     \
+  unsigned GENSYM(caml__max_spins) = (N);               \
+  for (; GENSYM(caml__spins) < GENSYM(caml__max_spins); \
+       cpu_relax(), ++GENSYM(caml__spins))
+
+/* Spin for unbounded time, this should only be used when there is a
+   very short critical section we are waiting on */
+#define SPIN_WAIT SPIN_WAIT_BACK_OFF(Max_spins_long)
+
+struct caml_plat_srcloc {
+  const char* file;
+  int line;
+  const char* function;
+};
+
+CAMLextern unsigned caml_plat_spin_wait(unsigned spins,
+                                        const struct caml_plat_srcloc* loc);
+
+Caml_inline unsigned caml_plat_spin_step(unsigned spins,
```

Comment:
In earlier versions of the code, this function had an extra branch and more complexity, and the ternary expression started looking even uglier than it already did. It's now equivalent to what the original `SPIN_WAIT` macro had, but I kept the refactoring because I think it's clearer.